ts/sagas/profile.ts

```diff
@@ -4,98 +4,98 @@
 import { call, Effect, put, select, takeLatest } from "redux-saga/effects";
 
 import {
-  ApiFetchResult,
-  ApiProfile,
-  fetchProfile,
-  isApiFetchFailure,
-  postProfile
-} from "../api";
+  BackendClient,
+  BasicResponseTypeWith401,
+  ProfileWithOrWithoutEmail
+} from "../api/backend";
+import { apiUrlPrefix } from "../config";
 import {
-  PROFILE_LOAD_FAILURE,
   PROFILE_LOAD_REQUEST,
-  PROFILE_LOAD_SUCCESS,
-  PROFILE_UPDATE_FAILURE,
-  PROFILE_UPDATE_REQUEST,
-  PROFILE_UPDATE_SUCCESS
+  PROFILE_UPSERT_REQUEST
 } from "../store/actions/constants";
-import { ProfileUpdateRequest } from "../store/actions/profile";
+import {
+  profileLoadFailure,
+  profileLoadSuccess,
+  profileUpsertFailure,
+  ProfileUpsertRequest,
+  profileUpsertSuccess
+} from "../store/actions/profile";
 import { sessionTokenSelector } from "../store/reducers/authentication";
 import { SessionToken } from "../types/SessionToken";
+import { callApiWith401ResponseStatusHandler } from "./api";
 
 // A saga to load the Profile.
 function* loadProfile(): Iterator<Effect> {
-  try {
-    // Get the token from the state
-    const sessionToken: SessionToken | undefined = yield select(
-      sessionTokenSelector
-    );
-    if (sessionToken === undefined) {
-      throw new Error("session token is not defined");
-    }
+  const sessionToken: SessionToken | undefined = yield select(
+    sessionTokenSelector
+  );
 
-    // Fetch the profile from the proxy
-    const response: ApiFetchResult<ApiProfile> = yield call(
-      fetchProfile,
-      sessionToken
+  if (sessionToken) {
+    const backendClient = BackendClient(apiUrlPrefix, sessionToken);
+
+    const response:
+      | BasicResponseTypeWith401<ProfileWithOrWithoutEmail>
+      | undefined = yield call(
+      callApiWith401ResponseStatusHandler,
+      backendClient.getProfile,
+      {}
     );
 
-    if (isApiFetchFailure(response)) {
-      // If the api response is an error then dispatch the PROFILE_LOAD_FAILURE action.
-      yield put({
-        type: PROFILE_LOAD_FAILURE,
-        payload: response.error
-      });
+    if (!response || response.status !== 200) {
+      // We got a error, send a SESSION_LOAD_FAILURE action
+      const error: Error = response ? response.value : Error();
+
+      yield put(profileLoadFailure(error));
     } else {
-      // If the api returns a valid Profile then dispatch the PROFILE_LOAD_SUCCESS action.
-      yield put({ type: PROFILE_LOAD_SUCCESS, payload: response.result });
+      // Ok we got a valid response, send a SESSION_LOAD_SUCCESS action
+      yield put(profileLoadSuccess(response.value));
     }
-  } catch (error) {
-    // If the api request raise an exception then dispatch the PROFILE_LOAD_FAILURE action.
-    yield put({ type: PROFILE_LOAD_FAILURE, payload: error });
+  } else {
+    // No SessionToken can't send the request
```

Comment:
I probably would make sense to disable any action in the UI that requires the session token until we have the session token - what do you think?

Reply:
Yes i was thinking in another story we can start (fork) the sagas that need the SessionToken after the event 'AUTHENTICATION_COMPLETED" passing the BackendClient. The only problem is that we need to "kill" and restart the sagas when the user is LoggedOut so we can pass the new BackedClient (with the new SessionToken).

Reply:
Reading your message again...when you say `disable action in the UI` you mean disable Buttons and so on?

Reply:
yes, why enable a button that triggers an action that you already know is going to fail?

until a valid session token is available, the app is basically offline

Reply:
...and the UI should behave accordingly (i.e. you can see the cached data but you can't do anything to update it)

Reply:
Added a story
https://www.pivotaltracker.com/story/show/158869140